### ISSUE TITLE
Make OSSL_STORE more flexible: support http: and relative local file: paths; improve documentation

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#14001]
Meanwhile a couple of apps are documented to support reading credentials (such as certificates and keys) not only from plain local files but also via some source specified as URI, e.g.,
```
[B<-in>      I<filename> |I<uri>]
[B<-trusted> I<filenames>|I<uris>]
[B<-signkey> I<filename> |I<uri>]
```
which is implemented via the OSSL_STORE API.
Some more explicit documentation is missing, also clarifying which URI types are supported for which credential types.

It looks like so far this is of pretty limited use, namely supporting only
* `file:` with absolute paths for various credential types
* `org.openssl.engine:` for keys

Would be good to support also
* relative `file:` paths (**update:** just for basic local system-dependent interpretation)
* `http:` with full paths (**update:** just using the basic HTTP support available)

for various credential types including certs, keys, and CRLs.

Apparently all building blocks needed for this are available and just would need to get connected.

**Comment by levitte** [comment]
It does support relative paths, when it's *just* a file name (no scheme).
When there is a URI scheme, there is no such thing as relative paths, at least as far as I read the specs.  Especially, this applies to `file:`, and I believe the same goes for `http:`

**Comment by levitte** [comment]
The architecture for OSSL_STORE is made so that additional schemes can be supported via engine or provider modules.

Re `http:`, is there a profile for certificates, CRLs etc these days?  I know that you have done some sort of client, but I haven't read the specs for how the URLs are constructed, or what else to expect, so I'm not confident about making an `http:` scheme implementation yet...

**Comment by DDvO** [comment]
> It does support relative paths, when it's _just_ a file name (no scheme).

Well, but then it does not bring anything new to apps regarding simple file names.

> When there is a URI scheme, there is no such thing as relative paths, at least as far as I read the specs. Especially, this applies to `file:`, and I believe the same goes for `http:`

Maybe there's no official scheme for relative local file (path) names, but I've seen support for this in the wild.

**Comment by levitte** [comment]
> > It does support relative paths, when it's _just_ a file name (no scheme).
> 
> Well, but then it does not bring anything new to apps regarding simple file names.

Not from a reference point of view, no.  But it's a scheme that we can support, and it's the default scheme (when the file is given without a scheme, the `file:` scheme is used under the hood, with the special property that relative paths are allowed)

> > When there is a URI scheme, there is no such thing as relative paths, at least as far as I read the specs. Especially, this applies to `file:`, and I believe the same goes for `http:`
> 
> Maybe there's no official scheme for relative local file (path) names, but I've seen support for this in the wild.

Sure, but I'd rather base our implementation on standards.

**Comment by richsalz** [comment]
You are correct that there is no relative path syntax for URI's; what you are seeing is the browser turning relative paths into absolute.

**Comment by DDvO** [comment]
> You are correct that there is no relative path syntax for URI's; what you are seeing is the browser turning relative paths into absolute.

Any handler for relative file path names in the end turns relative path names into absolute ones,
essentially by prepending it by 'the current path'.
The problem is just how to define 'the current path' for the given context.

**Comment by richsalz** [comment]
> The problem is just how to define 'the current path' for the given context.

Yup.  Then you get into things like symlinks (what does `..` mean), cross-platform portability, etc.

**Comment by DDvO** [comment]
I'd see the symlinks etc. issue very pragmatic: it simply depends on the local file system.
I see no need to make `file:` be interpretable in a cross-system manner,
(while the situation for relative/path names is of course different, and inherently much more complex, for `http:`).

**Comment by levitte** [comment]
Guys, before you go too far into the syntax rabbit hole, have a refresh read of [RFC 8089 (about the `file:` scheme)](https://tools.ietf.org/html/rfc8089) and [RFC 3986 (General URI syntax)](https://tools.ietf.org/html/rfc3986)

In short, the `file:` scheme specifies `path-absolute`, which is a unix-like path syntax.  VMS syntax *doesn't apply*

**Comment by DDvO** [comment]
> The architecture for OSSL_STORE is made so that additional schemes can be supported via engine or provider modules.

I see.

> Re `http:`, is there a profile for certificates, CRLs etc these days?

I don't know, but downloading certs and CRLs via HTTP is not uncommon, and OpenSSL has already support for this:
```
 X509 *X509_load_http(const char *url, BIO *bio, BIO *rbio, int timeout);
 X509_CRL *X509_CRL_load_http(const char *url, BIO *bio, BIO *rbio, int timeout);
 #define X509_http_nbio(rctx, pcert)
 #define X509_CRL_http_nbio(rctx, pcrl)
```
even if a profile is not officially defined.

> I know that you have done some sort of client, but I haven't read the specs for how the URLs are constructed, or what else to expect, so I'm not confident about making an `http:` scheme implementation yet...

The HTTP client is 'specified' as far as documented in `OSSL_HTTP_REQ_CTX.pod` and `OSSL_HTTP_transfer.pod`.
This includes a function `OSSL_HTTP_parse_url()`, which I've slightly extended from the original OCSP-centric one. It ...
```
parses its input string I<url> as a URL
of the form C<[http[s]://]address[:port][/path]> and splits it up into host,
port, and path components and a flag indicating whether it begins with 'https'.
The host component may be a DNS name or an IP address
where IPv6 addresses should be enclosed in square brackets C<[> and C<]>.
The port component is optional and defaults to "443" for HTTPS, else "80".
If the I<pport_num> argument is NULL the port specification
can be in mnemonic form such as "http" like with L<BIO_set_conn_port(3)>, else
it must be in numerical form and its integer value is assigned to I<*pport_num>.
The path component is also optional and defaults to "/".
```
Everything else regarding interpretation of the URL is left to the server.

**Comment by richsalz** [comment]
> Everything else regarding interpretation of the URL is left to the server.

Not stripping fragment ID's is non-compliant (https://example.com/foo#fragment) and should probably be fixed.

Not supporting name/password is a deficiency (https://example.com:443@[name]:[password]/foo) that probably should be fixed in the documentation for now.

**Comment by DDvO** [comment]
Good point that such URL portions should better be stripped and their non-support documented.
There also many many other HTTP features that the pretty basic client does not support, either.
I'm going to issue a feature request for one of these today: persistent connections.

**Comment by DDvO** [comment]
> Not stripping fragment ID's is non-compliant (https://example.com/foo#fragment) and should probably be fixed.
> 
> Not supporting name/password is a deficiency ([https://example.com:443@[name]:[password]/foo](https://example.com:443@%5Bname%5D:%5Bpassword%5D/foo)) that probably should be fixed in the documentation for now.

@richsalz, here is a PR just for you #14009 :wink: - well not quite, since you
asked for stripping `@[name]:[password]` from `https://example.com:443@[name]:[password]/foo`
but got stripping `name[:password]@` from `https://name[:password]@example.com:443/foo`

**Comment by nhorman** [comment]
Seems like we have a PR here that has been merged, which seemingly addresses the stated problem, so

Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
Closing as the issue was labeled as inactive for extended period of time and there was no additional input.
Please add a comment if you have any further information to support reopening the issue.
